### DISCUSSION TITLE
Release uploads fail too often

### DISCUSSION THREAD

**Discussion by lnicola** [community/community#9121] [Actions]
Hello! In our project, we use the GitHub API (well, Octokit) to upload release artifacts from a GitHub Action. I assume this is quite normal, however we the uploads fail very often with `ECONNRESET`.

Examples from today: https://github.com/rust-analyzer/rust-analyzer/actions/runs/1600722980/attempts/1, https://github.com/rust-analyzer/rust-analyzer/actions/runs/1600722980/attempts/2 (EDIT: it worked the third time). This has been happening pretty often, especially in (but not since) the past couple of months. Also note that we do retry on failure, but somehow one failure seems to be very likely followed by another one.

Our publishing step could indeed be improved. I tried that in https://github.com/lnicola/rust-analyzer/runs/4575266940?check_suite_focus=true, but notice how it kept retrying for 33 minutes. It somehow feels wrong for the API to fail this often.

I checked the `gh release upload` implementation and it doesn't seem to be doing anything special. It even retries faster and for fewer times. Maybe you have some idea about what might be happening?

---

Here's one example from the logs, in case they expire (please ignore the mis-formatted retry messages):

```text
HttpError: request to https://uploads.github.com/repos/lnicola/rust-analyzer/releases/55626630/assets?name=rust-analyzer-aarch64-pc-windows-msvc.gz& failed, reason: read ECONNRESET
RETRYING after {} s 1
HttpError: request to https://uploads.github.com/repos/lnicola/rust-analyzer/releases/55626630/assets?name=rust-analyzer-aarch64-pc-windows-msvc.gz& failed, reason: socket hang up
RETRYING after {} s 1
HttpError: request to https://uploads.github.com/repos/lnicola/rust-analyzer/releases/55626630/assets?name=rust-analyzer-aarch64-pc-windows-msvc.gz& failed, reason: socket hang up
RETRYING after {} s 4
HttpError: request to https://uploads.github.com/repos/lnicola/rust-analyzer/releases/55626630/assets?name=rust-analyzer-aarch64-pc-windows-msvc.gz& failed, reason: socket hang up
RETRYING after {} s 1
HttpError: request to https://uploads.github.com/repos/lnicola/rust-analyzer/releases/55626630/assets?name=rust-analyzer-aarch64-pc-windows-msvc.gz& failed, reason: socket hang up
RETRYING after {} s 11
HttpError: request to https://uploads.github.com/repos/lnicola/rust-analyzer/releases/55626630/assets?name=rust-analyzer-aarch64-pc-windows-msvc.gz& failed, reason: socket hang up
RETRYING after {} s 26
HttpError: request to https://uploads.github.com/repos/lnicola/rust-analyzer/releases/55626630/assets?name=rust-analyzer-aarch64-pc-windows-msvc.gz& failed, reason: socket hang up
RETRYING after {} s 31
HttpError: request to https://uploads.github.com/repos/lnicola/rust-analyzer/releases/55626630/assets?name=rust-analyzer-aarch64-pc-windows-msvc.gz& failed, reason: socket hang up
RETRYING after {} s 101
ERROR:  Validation Failed: {"resource":"ReleaseAsset","code":"already_exists","field":"name"}
HttpError: Validation Failed: {"resource":"ReleaseAsset","code":"already_exists","field":"name"}
```

Here the "duplicate file" error comes after a "failed" upload. That's not totally unexpected, but specific to my fork. The action we use deletes the release and restarts. In my fork I was trying to upload only the failed ones, but didn't delete them.

---

And [here](https://github.com/lnicola/rust-analyzer/runs/4610748269?check_suite_focus=true) is once with per-file retries:

![image](https://user-images.githubusercontent.com/308347/147146632-9fecdee7-95e4-45b7-8212-8ec932c6912a.png)

As you can see, it probably fixes the issue in practice, but it still failed four times within a single run.

**Comment by database64128** [comment]
Same here. In https://github.com/Shadowsocks-NET/QvStaticBuild/actions, nearly 2/3 of recent workflow runs failed because of this. For my other projects, about half of the release attempts were botched by this, and had to be re-run by hand.
